rbac/management/group/definer.py

```diff
@@ -186,19 +198,25 @@ def remove_roles(group, roles_or_role_ids, tenant, user=None):
     # This should not be necessary for system roles.
     custom_roles = roles.filter(tenant=tenant).select_for_update()
 
+    removed_roles = []
+    dual_write_handler = None
+    if group.tenant.tenant_name != "public":
+        dual_write_handler = RelationApiDualWriteGroupHandler(group, ReplicationEventType.UNASSIGN_ROLE)
+
     for policy in group.policies.all():
         for role in [*system_roles, *custom_roles]:
             # Only remove the role if it was attached
             if policy.roles.filter(pk=role.pk).exists():
                 policy.roles.remove(role)
                 logger.info(f"Removing role {role} from group {group.name} for tenant {tenant.org_id}.")
-
-                if tenant.tenant_name != "public":
-                    dual_write_handler = RelationApiDualWriteGroupHandler(group, ReplicationEventType.UNASSIGN_ROLE)
-                    dual_write_handler.replicate_removed_role(role)
-
+                if dual_write_handler:
+                    dual_write_handler.replicate_removed_roles([role])
                 # Send notifications
                 group_role_change_notification_handler(user, group, role, "removed")
+                removed_roles.append(role)
+
+    if dual_write_handler:
+        dual_write_handler.replicate()
```

Comment:
Similar to above, maybe like this:

```
    removed_roles = []

    for policy in group.policies.all():
        for role in [*system_roles, *custom_roles]:
            # Only remove the role if it was attached
            if policy.roles.filter(pk=role.pk).exists():
                policy.roles.remove(role)
                logger.info(f"Removing role {role} from group {group.name} for tenant {tenant.org_id}.")
                # Send notifications
                group_role_change_notification_handler(user, group, role, "removed")
                removed_roles.append(role)

    if group.tenant.tenant_name != "public":
        dual_write_handler = RelationApiDualWriteGroupHandler(group, ReplicationEventType.UNASSIGN_ROLE)
        dual_write_handler.replicate_removed_roles(removed_roles)
        dual_write_handler.replicate()
```

(and again either just `replicate_removed_roles`, or `remove_roles` then `replicate`)